File: Java/TelemetrySystem/src/main/java/tddmicroexercises/telemetrysystem/TelemetryDiagnosticControls.java
package tddmicroexercises.telemetrysystem;

public class TelemetryDiagnosticControls
{
    private final String DiagnosticChannelConnectionString = "*111#";
    
    private IConnection connectionControl;

    private IMessage messageControl;

    private String diagnosticInfo = "";

    public void setConnectionControl(IConnection connectionControl)
    {
        this.connectionControl = connectionControl;
    }

    public IConnection getConnectionControl()
    {
        return this.connectionControl;
    }

    public void setMessageControl(IMessage messageControl)
    {
        this.messageControl = messageControl;
    }

    public IMessage getMessageControl()
    {
        return this.messageControl;
    }

    public String getDiagnosticInfo(){
        return diagnosticInfo;
    }

    public void setDiagnosticInfo(String diagnosticInfo){
        this.diagnosticInfo = diagnosticInfo;
    }

    public void checkTransmission() throws Exception
    {
        diagnosticInfo = "";

        connectionControl.disconnect();

        int retryLeft = 3;
        while (connectionControl.getOnlineStatus() == false && retryLeft > 0)
        {
            connectionControl.connect(DiagnosticChannelConnectionString);
            retryLeft -= 1;
        }

        if(connectionControl.getOnlineStatus() == false)
        {
            throw new Exception("Unable to connect.");
        }

        messageControl.send(messageControl.DIAGNOSTIC_MESSAGE);
        diagnosticInfo = messageControl.receive();
    }
}

File: Javascript/telemetry-system/telemetry-diagnostic-controls.js

TelemetryDiagnosticControls = function (telemetryClient, telemetryData) {

	this._diagnosticChannelConnectionString = function () { return '*111#'; };
	this._diagnosticInfo = '';
	this._telemetryClient = telemetryClient;
	this._telemetryData = telemetryData;
};

TelemetryDiagnosticControls.Create = function () {
	return new TelemetryDiagnosticControls(new TelemetryClient(), new TelemetryData());
};

TelemetryDiagnosticControls.prototype = {

	readDiagnosticInfo: function() {
		return this._diagnosticInfo;
	},

	writeDiagnosticInfo: function(newValue) {
		this._diagnosticInfo = newValue;
	},

	checkTransmission: function() {

		this._diagnosticInfo = '';

		this._telemetryClient.disconnect();

		var retryLeft = 3;
		while (this._telemetryClient.onlineStatus() === false && retryLeft > 0) {
			this._telemetryClient.connect(this._diagnosticChannelConnectionString);
			retryLeft -= 1;
		}

		if (this._telemetryClient.onlineStatus() === false) {
			throw 'Unable to connect';
		}

		this._telemetryData.send(telemetryData.diagnosticMessage());
		this._diagnosticInfo = this._telemetryData.receive();
	}
};